File: src/lib/notifications.ts
import { supabase } from './supabase';
import { LINKEDIN_COMPANY_URL, SITE_URL } from '../config/site';
import { captureEmptyMutation, captureError } from './errorReporting';
import type { OutreachEmailTemplate } from '../types';

export type NotificationTemplate =
  | 'site_down'
  | 'badge_issued'
  | 'membership_welcome'
  | 'badge_request_received'
  | OutreachEmailTemplate;

/** Inbox charities should reply to. Also set as Resend reply_to on send. */
export const OUTREACH_REPLY_EMAIL = 'hello@ngoreality.com';

const OUTREACH_LINKEDIN_LINES = [
  `Follow us on LinkedIn:`,
  LINKEDIN_COMPANY_URL,
].join('\n');

/** NZ ownership / contact line on every outreach email. */
export const OUTREACH_NZ_FOOTER = [
  `—`,
  `NGOreality is New Zealand owned and operated.`,
  `Learn more:`,
  `https://www.ngoreality.com/public/about`,
  ``,
  OUTREACH_LINKEDIN_LINES,
  ``,
  `CEO: Al Baqshi`,
  `Email: ${OUTREACH_REPLY_EMAIL}`,
  `Phone: +64 27 338 8500`,
].join('\n');

function withLinkedInFollow(text: string): string {
  if (text.includes('linkedin.com/company/ngoreality')) return text;
  if (text.includes('https://www.ngoreality.com/public/about')) {
    return text.replace(
      'https://www.ngoreality.com/public/about',
      `https://www.ngoreality.com/public/about\n\n${OUTREACH_LINKEDIN_LINES}`,
    );
  }
  if (text.includes(`Email: ${OUTREACH_REPLY_EMAIL}`)) {
    return text.replace(
      `Email: ${OUTREACH_REPLY_EMAIL}`,
      `Email: ${OUTREACH_REPLY_EMAIL}\nLinkedIn: ${LINKEDIN_COMPANY_URL}`,
    );
  }
  return `${text}\n\n${OUTREACH_LINKEDIN_LINES}`;
}

/** Append the NZ footer unless the phone line is already present. */
export function withOutreachNzFooter(body: string): string {
  const text = (body ?? '').trimEnd();
  if (!text) return OUTREACH_NZ_FOOTER;
  if (text.includes(OUTREACH_REPLY_EMAIL)) {
    const withPhone = text.includes('+64 27 338 8500')
      ? text
      : `${text}\n\n${OUTREACH_NZ_FOOTER}`;
    return withLinkedInFollow(withPhone);
  }
  if (text.includes('Phone: +64 27 338 8500')) {
    return withLinkedInFollow(
      text.replace(
        'Phone: +64 27 338 8500',
        `Email: ${OUTREACH_REPLY_EMAIL}\nPhone: +64 27 338 8500`,
      ),
    );
  }
  if (text.includes('+64 27 338 8500')) return withLinkedInFollow(text);
  return withLinkedInFollow(`${text}\n\n${OUTREACH_NZ_FOOTER}`);
}

/** Absolute URL of an organisation's public directory page (`/public/org/<slug>`). */
export function publicProfileUrl(slug?: string | null): string | null {
  const clean = slug?.trim();
  if (!clean) return null;
  return `${SITE_URL}/public/org/${encodeURIComponent(clean)}`;
}

function buildMessage(
  template: NotificationTemplate,
  organizationName: string,
  extra?: Record<string, string>,
): { subject: string; body: string } {
  switch (template) {
    case 'site_down':
      return {
        subject: `[NGOreality] Website down — ${organizationName}`,
        body: [
          `Hello,`,
          ``,
          `Our monitors detected that the website for ${organizationName} is not responding.`,
          ``,
          `As an NGOreality member you receive this alert automatically. If you need hands-on help fixing the issue, reply to this email — support is billed separately from your annual membership.`,
          ``,
          `— NGOreality monitoring`,
          `Follow us on LinkedIn: ${LINKEDIN_COMPANY_URL}`,
        ].join('\n'),
      };
    case 'badge_issued':
      return {
        subject: `[NGOreality] Your Reality Badge is active — ${organizationName}`,
        body: [
          `Congratulations — ${organizationName} has met NGOreality public trust standards and your Reality Badge is now active.`,
          extra?.verificationId ? `Badge ID: ${extra.verificationId}` : '',
          ``,
          `Your membership includes website monitoring for one year. We will email you if your site goes down.`,
          ``,
          `— NGOreality`,
          `Follow us on LinkedIn: ${LINKEDIN_COMPANY_URL}`,
        ]
          .filter(Boolean)
          .join('\n'),
      };
    case 'membership_welcome':
      return {
        subject: `[NGOreality] Membership active — ${organizationName}`,
        body: [
          `Thank you — annual NGOreality membership for ${organizationName} is now active.`,
          ``,
          `Included for one year:`,
          `• Reality Badge (after standards review)`,
          `• Website uptime monitoring and email alerts`,
          ``,
          `Sign in to your portal for member-only security checklist items (repository, security baseline, and more).`,
          ``,
          `Consulting, custom sites, and hands-on support are available separately.`,
          ``,
          `— NGOreality`,
          `Follow us on LinkedIn: ${LINKEDIN_COMPANY_URL}`,
        ].join('\n'),
      };
    case 'badge_request_received': {
      const ref = extra?.paymentReference ?? 'NGR-…';
      return {
        subject: `[NGOreality] Application received — ${organizationName}`,
        body: [
          `Kia ora,`,
          ``,
          `Thank you for submitting your NGOreality verification request for ${organizationName}.`,
          ``,
          `We will review your organisation against our public trust standards and contact you if anything needs to be updated on your site.`,
          ``,
          `Typical timeline: about 5–10 business days after payment and standards are in order. Your Reality Badge is issued when standards pass and membership is active.`,
          ``,
          `Bank transfer — use this payment reference exactly: ${ref}`,
          `Amount: NZD $70.00 annual membership`,
          `Allow up to 3 business days after we receive your transfer for it to be applied.`,
          ``,
          `Paymark (Online EFTPOS) and Airwallex payments are coming soon to the portal.`,
          ``,
          `— NGOreality`,
          `Follow us on LinkedIn: ${LINKEDIN_COMPANY_URL}`,
        ].join('\n'),
      };
    }
    case 'outreach_cold_invite':
      return {
        subject: `[NGOreality] A note for ${organizationName}`,
        body: [
          `Kia ora ${organizationName},`,
          ``,
          `We came across ${organizationName} on the New Zealand Charities Register and wanted to introduce NGOreality.`,
          ``,
          `Your organisation already has a public presence online, so our suggestion isn't about replacing what you have.`,
          ``,
          `NGOreality gives New Zealand charities another way to demonstrate their organisation's public information and trust standards through a verified Reality Badge and profile.`,
          ``,
          `You can claim and onboard your organisation here:`,
          `{page}`,
          ``,
          `Reality Badge membership is NZD $70 per year and includes the badge and website monitoring.`,
          ``,
          `If you'd like to know whether it would be useful for ${organizationName}, simply reply to this email and we'll explain it.`,
          ``,
          `—`,
          `NGOreality is New Zealand owned and operated.`,
          ``,
          `Learn more:`,
          `https://www.ngoreality.com/public/about`,
          ``,
          `Follow us on LinkedIn:`,
          LINKEDIN_COMPANY_URL,
          ``,
          `CEO: Al Baqshi`,
          `Email: ${OUTREACH_REPLY_EMAIL}`,
          `Phone: +64 27 338 8500`,
        ].join('\n'),
      };
    case 'outreach_no_website':
      return {
        subject: `[NGOreality] A note for ${organizationName}`,
        body: [
          `Kia ora ${organizationName},`,
          ``,
          `We came across ${organizationName} while looking at New Zealand charities and noticed there isn't a public website listed.`,
          ``,
          `We're not suggesting you need an expensive new website.`,
          ``,
          `Sometimes a clear public profile, updated organisation information, and a simple trust-focused page are enough to make it much easier for people to understand who you are and what your organisation does.`,
          ``,
          `If that would help ${organizationName}, a trust landing page with us is NZD $650, one-off. We build it at that fixed price.`,
          ``,
          `You can claim your NGOreality profile here:`,
          `{page}`,
          ``,
          `If you'd like, reply to this email and we can explain what we would recommend for ${organizationName} and what it would cost.`,
          ``,
          `No pressure — we're happy to have a conversation first.`,
          ``,
          `—`,
          `NGOreality is New Zealand owned and operated.`,
          ``,
          `Learn more:`,
          `https://www.ngoreality.com/public/about`,
          ``,
          `Follow us on LinkedIn:`,
          LINKEDIN_COMPANY_URL,
          ``,
          `CEO: Al Baqshi`,
          `Email: ${OUTREACH_REPLY_EMAIL}`,
          `Phone: +64 27 338 8500`,
        ].join('\n'),
      };
    case 'outreach_website_help':
      return {
        subject: `[NGOreality] A note for ${organizationName}`,
        body: [
          `Kia ora ${organizationName},`,
          ``,
          `We came across ${organizationName} while looking at New Zealand charities and noticed your website does not appear to be loading at the moment.`,
          ``,
          `That happens more often than people think, and it is not a criticism. When a site is down, it is simply harder for people to find you and understand the work you do.`,
          ``,
          `We're not suggesting you need an expensive new website.`,
          ``,
          `Sometimes a clear public profile and a simple trust-focused page are enough — and if you would like us to build that for ${organizationName}, a trust landing page is NZD $650, one-off. We build it at that fixed price.`,
          ``,
          `You can claim your NGOreality profile here:`,
          `{page}`,
          ``,
          `If you'd like, reply to this email and we can explain what we would recommend for ${organizationName} and what it would cost.`,
          ``,
          `No pressure — we're happy to have a conversation first.`,
          ``,
          `—`,
          `NGOreality is New Zealand owned and operated.`,
          ``,
          `Learn more:`,
          `https://www.ngoreality.com/public/about`,
          ``,
          `Follow us on LinkedIn:`,
          LINKEDIN_COMPANY_URL,
          ``,
          `CEO: Al Baqshi`,
          `Email: ${OUTREACH_REPLY_EMAIL}`,
          `Phone: +64 27 338 8500`,
        ].join('\n'),
      };
    default:
      return { subject: 'NGOreality', body: '' };
  }
}

/**
 * Replace placeholders when staff edits the draft before send.
 *  - `{name}` / `{organizationName}` → the organisation's name
 *  - `{page}` / `{profileUrl}`       → link to its public directory page (/public/org/<slug>)
 * The same brackets are honoured server-side by `outreach_enqueue_emails` (migration 064).
 */
export function personalizeOutreachDraft(
  text: string,
  organizationName: string,
  profileUrl?: string | null,
): string {
  const page = profileUrl ?? '';
  return text
    .replace(/\{name\}/gi, organizationName)
    .replace(/\{organizationName\}/gi, organizationName)
    .replace(/\{page\}/gi, page)
    .replace(/\{profileUrl\}/gi, page);
}

/** Queue an email for the worker / manual send from CRM. */
export async function queueNotification(input: {
  organizationId: string;
  template: NotificationTemplate;
  recipientEmail: string;
  organizationName: string;
  /** Public directory slug — lets `{page}` in a staff draft resolve to the org's profile link. */
  organizationSlug?: string | null;
  incidentId?: string;
  extra?: Record<string, string>;
  subjectOverride?: string;
  bodyOverride?: string;
}): Promise<{ error: string | null }> {
  const email = input.recipientEmail.trim();
  if (!email) return { error: 'No recipient email' };

  const built = buildMessage(input.template, input.organizationName, {
    ...input.extra,
    organizationId: input.organizationId,
  });
  const profileUrl = publicProfileUrl(input.organizationSlug);
  const subject = personalizeOutreachDraft(
    input.subjectOverride ?? built.subject,
    input.organizationName,
    profileUrl,
  );
  const body = personalizeOutreachDraft(
    input.bodyOverride ?? built.body,
    input.organizationName,
    profileUrl,
  );

  const { data, error } = await supabase.from('notification_events').insert({
    organization_id: input.organizationId,
    incident_id: input.incidentId ?? null,
    template: input.template,
    recipient_email: email,
    subject,
    body_text: body,
    status: 'pending',
  }).select('id');

  if (error) {
    return { error: captureError(error, { where: 'queueNotification' }) };
  }
  if (!data?.length) {
    return { error: captureEmptyMutation('queueNotification.empty', 'The email was not queued. Refresh and try again.') };
  }
  return { error: null };
}

/** After queueing, try immediate send via Go API when configured. */
export async function queueAndTrySend(input: Parameters<typeof queueNotification>[0]): Promise<{
  error: string | null;
  flushError: string | null;
}> {
  const { error } = await queueNotification(input);
  if (error) return { error, flushError: null };

  try {
    const { flushPendingNotifications } = await import('./monitorApi');
    await flushPendingNotifications();
    return { error: null, flushError: null };
  } catch (e) {
    captureError(e, { where: 'queueAndTrySend.flush' });
    return {
      error: null,
      flushError: e instanceof Error ? e.message : 'Could not flush notifications',
    };
  }
}

export function previewOutreachEmail(
  template: OutreachEmailTemplate,
  organizationName: string,
  extra?: Record<string, string>,
) {
  return buildMessage(template, organizationName, extra);
}

File: backend/internal/store/notifications.go
package store

import (
	"context"
	"fmt"
	"strings"
	"time"
)

type NotificationEvent struct {
	ID               string
	OrganizationID   string
	IncidentID       *string
	Template         string
	RecipientEmail   string
	Subject          string
	BodyText         string
	Status           string
	ErrorMessage     string
	CreatedAt        time.Time
	OrganizationName string
}

func (s *Store) HasActiveMembership(ctx context.Context, orgID string) (bool, error) {
	var ok bool
	err := s.pool.QueryRow(ctx, `SELECT public.has_active_membership($1)`, orgID).Scan(&ok)
	return ok, err
}

func (s *Store) OrganizationAlertTarget(ctx context.Context, orgID string) (email, orgName string, err error) {
	err = s.pool.QueryRow(ctx, `
		SELECT
		  coalesce(nullif(trim(o.email), ''), nullif(trim(c.email), ''), ''),
		  o.name
		FROM organizations o
		LEFT JOIN LATERAL (
		  SELECT email FROM contacts
		  WHERE organization_id = o.id AND is_primary = true
		  ORDER BY created_at
		  LIMIT 1
		) c ON true
		WHERE o.id = $1
	`, orgID).Scan(&email, &orgName)
	if err != nil {
		return "", "", err
	}
	email = strings.TrimSpace(email)
	orgName = strings.TrimSpace(orgName)
	if email == "" {
		return "", orgName, fmt.Errorf("no alert email for organization")
	}
	return email, orgName, nil
}

func (s *Store) HasPendingSiteDownNotification(ctx context.Context, orgID string) (bool, error) {
	var n int
	err := s.pool.QueryRow(ctx, `
		SELECT count(*)::int FROM notification_events
		WHERE organization_id = $1
		  AND template = 'site_down'
		  AND status IN ('pending', 'sent')
		  AND created_at > now() - interval '24 hours'
	`, orgID).Scan(&n)
	return n > 0, err
}

func (s *Store) InsertNotificationEvent(ctx context.Context, orgID string, incidentID *string, template, recipientEmail, subject, body string) error {
	_, err := s.pool.Exec(ctx, `
		INSERT INTO notification_events (
		  organization_id, incident_id, template, recipient_email, subject, body_text, status
		) VALUES ($1, $2, $3, $4, $5, $6, 'pending')
	`, orgID, incidentID, template, recipientEmail, subject, body)
	return err
}

func siteDownEmail(orgName string) (subject, body string) {
	subject = fmt.Sprintf("[NGOreality] Website down — %s", orgName)
	body = strings.Join([]string{
		"Hello,",
		"",
		fmt.Sprintf("Our monitors detected that the website for %s is not responding.", orgName),
		"",
		"As an NGOreality member you receive this alert automatically. If you need hands-on help fixing the issue, reply to this email — support is billed separately from your annual membership.",
		"",
		"— NGOreality monitoring",
	}, "\n")
	return subject, body
}

// QueueSiteDownAlert enqueues a down alert for paying members (deduped per 24h).
func (s *Store) QueueSiteDownAlert(ctx context.Context, orgID, incidentID string) error {
	active, err := s.HasActiveMembership(ctx, orgID)
	if err != nil || !active {
		return err
	}
	dup, err := s.HasPendingSiteDownNotification(ctx, orgID)
	if err != nil || dup {
		return err
	}
	email, orgName, err := s.OrganizationAlertTarget(ctx, orgID)
	if err != nil {
		return err
	}
	subject, body := siteDownEmail(orgName)
	inc := incidentID
	return s.InsertNotificationEvent(ctx, orgID, &inc, "site_down", email, subject, body)
}

func (s *Store) ListPendingNotifications(ctx context.Context, limit int) ([]NotificationEvent, error) {
	if limit <= 0 {
		limit = 50
	}
	rows, err := s.pool.Query(ctx, `
		SELECT
		  n.id, n.organization_id, n.incident_id, n.template,
		  n.recipient_email, n.subject, n.body_text, n.status,
		  coalesce(n.error_message, ''), n.created_at,
		  coalesce(o.name, '')
		FROM notification_events n
		LEFT JOIN organizations o ON o.id = n.organization_id
		WHERE n.status = 'pending'
		ORDER BY n.created_at ASC
		LIMIT $1
	`, limit)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var out []NotificationEvent
	for rows.Next() {
		var e NotificationEvent
		var incidentID *string
		if err := rows.Scan(
			&e.ID, &e.OrganizationID, &incidentID, &e.Template,
			&e.RecipientEmail, &e.Subject, &e.BodyText, &e.Status,
			&e.ErrorMessage, &e.CreatedAt, &e.OrganizationName,
		); err != nil {
			return nil, err
		}
		e.IncidentID = incidentID
		out = append(out, e)
	}
	return out, rows.Err()
}

func (s *Store) MarkNotificationSent(ctx context.Context, id string, incidentID *string) error {
	tx, err := s.pool.Begin(ctx)
	if err != nil {
		return err
	}
	defer tx.Rollback(ctx)

	_, err = tx.Exec(ctx, `
		UPDATE notification_events
		SET status = 'sent', sent_at = now(), error_message = ''
		WHERE id = $1
	`, id)
	if err != nil {
		return err
	}
	if incidentID != nil && *incidentID != "" {
		_, err = tx.Exec(ctx, `
			UPDATE website_incidents
			SET org_notified_at = now()
			WHERE id = $1 AND org_notified_at IS NULL
		`, *incidentID)
		if err != nil {
			return err
		}
	}
	return tx.Commit(ctx)
}

func (s *Store) MarkNotificationFailed(ctx context.Context, id, errMsg string) error {
	_, err := s.pool.Exec(ctx, `
		UPDATE notification_events
		SET status = 'failed', error_message = $2
		WHERE id = $1
	`, id, errMsg)
	return err
}

func (s *Store) RequeueNotification(ctx context.Context, id string) error {
	_, err := s.pool.Exec(ctx, `
		UPDATE notification_events
		SET status = 'pending', error_message = '', sent_at = NULL
		WHERE id = $1 AND status = 'failed'
	`, id)
	return err
}

// NotificationSummary counts by status for CRM dashboard.
type NotificationSummary struct {
	Pending int `json:"pending"`
	Sent    int `json:"sent"`
	Failed  int `json:"failed"`
}

func (s *Store) NotificationSummary(ctx context.Context) (NotificationSummary, error) {
	var sum NotificationSummary
	err := s.pool.QueryRow(ctx, `
		SELECT
		  count(*) FILTER (WHERE status = 'pending')::int,
		  count(*) FILTER (WHERE status = 'sent')::int,
		  count(*) FILTER (WHERE status = 'failed')::int
		FROM notification_events
		WHERE created_at > now() - interval '30 days'
	`).Scan(&sum.Pending, &sum.Sent, &sum.Failed)
	return sum, err
}

// ListRecentNotifications for CRM (all statuses).
func (s *Store) ListRecentNotifications(ctx context.Context, limit int) ([]NotificationEvent, error) {
	if limit <= 0 {
		limit = 100
	}
	rows, err := s.pool.Query(ctx, `
		SELECT
		  n.id, n.organization_id, n.incident_id, n.template,
		  n.recipient_email, n.subject, n.body_text, n.status,
		  coalesce(n.error_message, ''), n.created_at,
		  coalesce(o.name, '')
		FROM notification_events n
		LEFT JOIN organizations o ON o.id = n.organization_id
		ORDER BY n.created_at DESC
		LIMIT $1
	`, limit)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var out []NotificationEvent
	for rows.Next() {
		var e NotificationEvent
		var incidentID *string
		if err := rows.Scan(
			&e.ID, &e.OrganizationID, &incidentID, &e.Template,
			&e.RecipientEmail, &e.Subject, &e.BodyText, &e.Status,
			&e.ErrorMessage, &e.CreatedAt, &e.OrganizationName,
		); err != nil {
			return nil, err
		}
		e.IncidentID = incidentID
		out = append(out, e)
	}
	return out, rows.Err()
}

// OpenIncidentsWithoutOrgNotify lists open incidents where member org was not emailed yet.
func (s *Store) OpenIncidentsNeedingNotify(ctx context.Context, limit int) ([]struct {
	IncidentID     string
	OrganizationID string
}, error) {
	rows, err := s.pool.Query(ctx, `
		SELECT i.id, i.organization_id
		FROM website_incidents i
		WHERE i.closed_at IS NULL
		  AND i.org_notified_at IS NULL
		  AND public.has_active_membership(i.organization_id)
		ORDER BY i.opened_at ASC
		LIMIT $1
	`, limit)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var out []struct {
		IncidentID     string
		OrganizationID string
	}
	for rows.Next() {
		var row struct {
			IncidentID     string
			OrganizationID string
		}
		if err := rows.Scan(&row.IncidentID, &row.OrganizationID); err != nil {
			return nil, err
		}
		out = append(out, row)
	}
	return out, rows.Err()
}
